brainflow-dev/brainflow
# restoring openmp

Signed-off-by: Andrey Parfenov <a1994ndrey@gmail.com>

Idea of this PR:

* OpenMP provides much better perf than C++ 11 threads, check #162 
* There are problems with linking related to OpenMP, check #86 
* To fix these problems for Windows and MacOS I bundle vcomp.dll and libomp.dylib with distributed packages and try to preload them and\or set env variables to find these libs. For linux seems like its possible to link it statically with some tricks
* Still need to test that it works on machines wo openmp installed(or without c++ redist in case of windows)

Tools which I used to check dependencies:

* Dependency Walker on Windows
* ldd on linux
* ottol -l on MacOS

How to test:
* for linux run ldd for precompiled library(libDataHandler.so) and check that there is no libgomp and check that it works
* for Windows on machine wo c++ redist run dependency walker for DataHandler.dll and check that vcomp140.dll is not found, run python and java code and confirm that it works
* for MacOS on machine wo libomp installed run otool -L and check that libomp is not found, run python and java tests using precompiled libs with openmp and confirm that it works

Also, need to test perf for all methods using OMP_NUM_THREADS env variable and python band_power_all script
